File: open_source_projects/Deepbox-main/benchmarks/python/10_preprocess.py
"""
Benchmark 10 — Preprocessing
scikit-learn
"""

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

import numpy as np
from sklearn.preprocessing import (
    StandardScaler, MinMaxScaler, RobustScaler, MaxAbsScaler,
    Normalizer, PowerTransformer, QuantileTransformer,
    LabelEncoder, OneHotEncoder, OrdinalEncoder, LabelBinarizer,
)
from sklearn.model_selection import (
    train_test_split, KFold, StratifiedKFold, GroupKFold, LeaveOneOut, LeavePOut,
)
from utils import run, create_suite, header, footer

suite = create_suite("preprocess", "scikit-learn")
header("Benchmark 10 — Preprocessing", "scikit-learn")

# ── Data generators ──────────────────────────────────────

rng = np.random.RandomState(42)

X200 = rng.randn(200, 5) * 50
X500 = rng.randn(500, 10) * 50
X1k = rng.randn(1000, 10) * 50
X5k = rng.randn(5000, 10) * 50
Xpos500 = np.abs(rng.randn(500, 10)) * 50 + 1
Xpos1k = np.abs(rng.randn(1000, 10)) * 50 + 1
y200 = rng.randint(0, 3, 200)
y500 = rng.randint(0, 5, 500)
y1k = rng.randint(0, 5, 1000)

# ── StandardScaler ──────────────────────────────────────

run(suite, "StandardScaler fit", "200x5", lambda: StandardScaler().fit(X200))
run(suite, "StandardScaler fit", "500x10", lambda: StandardScaler().fit(X500))
run(suite, "StandardScaler fit", "1Kx10", lambda: StandardScaler().fit(X1k))
ss = StandardScaler().fit(X500)
run(suite, "StandardScaler transform", "500x10", lambda: ss.transform(X500))
run(suite, "StandardScaler transform", "1Kx10", lambda: ss.transform(X1k))
run(suite, "StandardScaler fit+transform", "5Kx10", lambda: StandardScaler().fit_transform(X5k))

# ── MinMaxScaler ────────────────────────────────────────

run(suite, "MinMaxScaler fit", "200x5", lambda: MinMaxScaler().fit(X200))
run(suite, "MinMaxScaler fit", "500x10", lambda: MinMaxScaler().fit(X500))
run(suite, "MinMaxScaler fit", "1Kx10", lambda: MinMaxScaler().fit(X1k))
mms = MinMaxScaler().fit(X500)
run(suite, "MinMaxScaler transform", "500x10", lambda: mms.transform(X500))
run(suite, "MinMaxScaler transform", "1Kx10", lambda: mms.transform(X1k))

# ── RobustScaler ────────────────────────────────────────

run(suite, "RobustScaler fit", "500x10", lambda: RobustScaler().fit(X500))
run(suite, "RobustScaler fit", "1Kx10", lambda: RobustScaler().fit(X1k))
rs = RobustScaler().fit(X500)
run(suite, "RobustScaler transform", "500x10", lambda: rs.transform(X500))

# ── MaxAbsScaler ────────────────────────────────────────

run(suite, "MaxAbsScaler fit", "500x10", lambda: MaxAbsScaler().fit(X500))
run(suite, "MaxAbsScaler fit", "1Kx10", lambda: MaxAbsScaler().fit(X1k))
mas = MaxAbsScaler().fit(X500)
run(suite, "MaxAbsScaler transform", "500x10", lambda: mas.transform(X500))

# ── Normalizer ──────────────────────────────────────────

run(suite, "Normalizer fit+transform", "500x10", lambda: Normalizer().fit_transform(X500))
run(suite, "Normalizer fit+transform", "1Kx10", lambda: Normalizer().fit_transform(X1k))

# ── PowerTransformer ────────────────────────────────────

run(suite, "PowerTransformer fit", "500x10", lambda: PowerTransformer(method="yeo-johnson").fit(Xpos500))
pt = PowerTransformer(method="yeo-johnson").fit(Xpos500)
run(suite, "PowerTransformer transform", "500x10", lambda: pt.transform(Xpos500))

# ── QuantileTransformer ─────────────────────────────────

run(suite, "QuantileTransformer fit", "500x10", lambda: QuantileTransformer().fit(X500))
qt = QuantileTransformer().fit(X500)
run(suite, "QuantileTransformer transform", "500x10", lambda: qt.transform(X500))

# ── LabelEncoder ────────────────────────────────────────

cats = ["cat", "dog", "fish", "bird", "snake"]
str500 = np.array([cats[i % 5] for i in range(500)])
str1k = np.array([cats[i % 5] for i in range(1000)])

run(suite, "LabelEncoder fit", "500 labels", lambda: LabelEncoder().fit(str500))
run(suite, "LabelEncoder fit", "1K labels", lambda: LabelEncoder().fit(str1k))
le = LabelEncoder().fit(str500)
run(suite, "LabelEncoder transform", "500 labels", lambda: le.transform(str500))
run(suite, "LabelEncoder transform", "1K labels", lambda: le.transform(str1k))

# ── OneHotEncoder ───────────────────────────────────────

str500_2d = str500.reshape(-1, 1)
str1k_2d = str1k.reshape(-1, 1)

run(suite, "OneHotEncoder fit", "500 samples", lambda: OneHotEncoder(sparse_output=False).fit(str500_2d))
run(suite, "OneHotEncoder fit", "1K samples", lambda: OneHotEncoder(sparse_output=False).fit(str1k_2d))
ohe = OneHotEncoder(sparse_output=False).fit(str500_2d)
run(suite, "OneHotEncoder transform", "500 samples", lambda: ohe.transform(str500_2d))

# ── OrdinalEncoder ──────────────────────────────────────

run(suite, "OrdinalEncoder fit", "500 samples", lambda: OrdinalEncoder().fit(str500_2d))
oe = OrdinalEncoder().fit(str500_2d)
run(suite, "OrdinalEncoder transform", "500 samples", lambda: oe.transform(str500_2d))

# ── LabelBinarizer ──────────────────────────────────────

run(suite, "LabelBinarizer fit", "500 samples", lambda: LabelBinarizer().fit(str500))
lbin = LabelBinarizer().fit(str500)
run(suite, "LabelBinarizer transform", "500 samples", lambda: lbin.transform(str500))

# ── trainTestSplit ──────────────────────────────────────

run(suite, "trainTestSplit", "200x5", lambda: train_test_split(X200, y200, test_size=0.2))
run(suite, "trainTestSplit", "500x10", lambda: train_test_split(X500, y500, test_size=0.2))
run(suite, "trainTestSplit", "1Kx10", lambda: train_test_split(X1k, y1k, test_size=0.2))

# ── KFold ───────────────────────────────────────────────

def kfold_iter(n_splits, X):
    kf = KFold(n_splits=n_splits)
    for _ in kf.split(X): pass

run(suite, "KFold (k=5)", "500 samples", lambda: kfold_iter(5, X500))
run(suite, "KFold (k=5)", "1K samples", lambda: kfold_iter(5, X1k))
run(suite, "KFold (k=10)", "1K samples", lambda: kfold_iter(10, X1k))

# ── StratifiedKFold ─────────────────────────────────────

def skfold_iter(n_splits, X, y):
    sf = StratifiedKFold(n_splits=n_splits)
    for _ in sf.split(X, y): pass

run(suite, "StratifiedKFold (k=5)", "500 samples", lambda: skfold_iter(5, X500, y500))
run(suite, "StratifiedKFold (k=5)", "1K samples", lambda: skfold_iter(5, X1k, y1k))

# ── LeaveOneOut ─────────────────────────────────────────

Xsmall = rng.randn(50, 3)

def loo_iter():
    loo = LeaveOneOut()
    for _ in loo.split(Xsmall): pass

run(suite, "LeaveOneOut", "50 samples", loo_iter)

footer(suite, "sklearn-preprocess.json")

File: open_source_projects/Deepbox-main/benchmarks/deepbox/10-preprocess.ts
/**
 * Benchmark 10 — Preprocessing
 * Deepbox vs scikit-learn
 */

import { tensor } from "deepbox/ndarray";
import {
	KFold,
	LabelBinarizer,
	LabelEncoder,
	LeaveOneOut,
	MaxAbsScaler,
	MinMaxScaler,
	Normalizer,
	OneHotEncoder,
	OrdinalEncoder,
	PowerTransformer,
	QuantileTransformer,
	RobustScaler,
	StandardScaler,
	StratifiedKFold,
	trainTestSplit,
} from "deepbox/preprocess";
import { createSuite, footer, header, run } from "../utils";

const suite = createSuite("preprocess");
header("Benchmark 10 — Preprocessing");

// ── Data generators ──────────────────────────────────────

function seededRng(seed: number) {
	let s = seed >>> 0;
	return () => {
		s = (s * 1664525 + 1013904223) >>> 0;
		return s / 2 ** 32;
	};
}

function makeNumeric(n: number, f: number, seed: number) {
	const rand = seededRng(seed);
	const data: number[][] = [];
	for (let i = 0; i < n; i++) {
		const row: number[] = [];
		for (let j = 0; j < f; j++) row.push(rand() * 100 - 50);
		data.push(row);
	}
	return tensor(data);
}

function makePositive(n: number, f: number, seed: number) {
	const rand = seededRng(seed);
	const data: number[][] = [];
	for (let i = 0; i < n; i++) {
		const row: number[] = [];
		for (let j = 0; j < f; j++) row.push(rand() * 100 + 1);
		data.push(row);
	}
	return tensor(data);
}

const X200 = makeNumeric(200, 5, 42);
const X500 = makeNumeric(500, 10, 42);
const X1k = makeNumeric(1000, 10, 42);
const X5k = makeNumeric(5000, 10, 42);
const Xpos500 = makePositive(500, 10, 42);
const _Xpos1k = makePositive(1000, 10, 42);

function makeLabels(n: number, k: number, seed: number) {
	const rand = seededRng(seed);
	const labels: number[] = [];
	for (let i = 0; i < n; i++) labels.push(Math.floor(rand() * k));
	return tensor(labels);
}

const y200 = makeLabels(200, 3, 42);
const y500 = makeLabels(500, 5, 42);
const y1k = makeLabels(1000, 5, 42);

// ── StandardScaler ──────────────────────────────────────

run(suite, "StandardScaler fit", "200x5", () => new StandardScaler().fit(X200));
run(suite, "StandardScaler fit", "500x10", () => new StandardScaler().fit(X500));
run(suite, "StandardScaler fit", "1Kx10", () => new StandardScaler().fit(X1k));
const ss = new StandardScaler().fit(X500);
run(suite, "StandardScaler transform", "500x10", () => ss.transform(X500));
run(suite, "StandardScaler transform", "1Kx10", () => ss.transform(X1k));
run(suite, "StandardScaler fit+transform", "5Kx10", () =>
	new StandardScaler().fit(X5k).transform(X5k)
);

// ── MinMaxScaler ────────────────────────────────────────

run(suite, "MinMaxScaler fit", "200x5", () => new MinMaxScaler().fit(X200));
run(suite, "MinMaxScaler fit", "500x10", () => new MinMaxScaler().fit(X500));
run(suite, "MinMaxScaler fit", "1Kx10", () => new MinMaxScaler().fit(X1k));
const mms = new MinMaxScaler().fit(X500);
run(suite, "MinMaxScaler transform", "500x10", () => mms.transform(X500));
run(suite, "MinMaxScaler transform", "1Kx10", () => mms.transform(X1k));

// ── RobustScaler ────────────────────────────────────────

run(suite, "RobustScaler fit", "500x10", () => new RobustScaler().fit(X500));
run(suite, "RobustScaler fit", "1Kx10", () => new RobustScaler().fit(X1k));
const rs = new RobustScaler().fit(X500);
run(suite, "RobustScaler transform", "500x10", () => rs.transform(X500));

// ── MaxAbsScaler ────────────────────────────────────────

run(suite, "MaxAbsScaler fit", "500x10", () => new MaxAbsScaler().fit(X500));
run(suite, "MaxAbsScaler fit", "1Kx10", () => new MaxAbsScaler().fit(X1k));
const mas = new MaxAbsScaler().fit(X500);
run(suite, "MaxAbsScaler transform", "500x10", () => mas.transform(X500));

// ── Normalizer ──────────────────────────────────────────

run(suite, "Normalizer fit+transform", "500x10", () => new Normalizer().fit(X500).transform(X500));
run(suite, "Normalizer fit+transform", "1Kx10", () => new Normalizer().fit(X1k).transform(X1k));

// ── PowerTransformer ────────────────────────────────────

run(suite, "PowerTransformer fit", "500x10", () => new PowerTransformer().fit(Xpos500));
const pt = new PowerTransformer().fit(Xpos500);
run(suite, "PowerTransformer transform", "500x10", () => pt.transform(Xpos500));

// ── QuantileTransformer ─────────────────────────────────

run(suite, "QuantileTransformer fit", "500x10", () => new QuantileTransformer().fit(X500));
const qt = new QuantileTransformer().fit(X500);
run(suite, "QuantileTransformer transform", "500x10", () => qt.transform(X500));

// ── LabelEncoder ────────────────────────────────────────

const strLabels500 = tensor(
	Array.from({ length: 500 }, (_, i) => ["cat", "dog", "fish", "bird", "snake"][i % 5])
);
const strLabels1k = tensor(
	Array.from({ length: 1000 }, (_, i) => ["cat", "dog", "fish", "bird", "snake"][i % 5])
);

run(suite, "LabelEncoder fit", "500 labels", () => new LabelEncoder().fit(strLabels500));
run(suite, "LabelEncoder fit", "1K labels", () => new LabelEncoder().fit(strLabels1k));
const le = new LabelEncoder().fit(strLabels500);
run(suite, "LabelEncoder transform", "500 labels", () => le.transform(strLabels500));
run(suite, "LabelEncoder transform", "1K labels", () => le.transform(strLabels1k));

// ── OneHotEncoder ───────────────────────────────────────

const strLabels500_2d = strLabels500.reshape([500, 1]);
const strLabels1k_2d = strLabels1k.reshape([1000, 1]);

run(suite, "OneHotEncoder fit", "500 samples", () => new OneHotEncoder().fit(strLabels500_2d));
run(suite, "OneHotEncoder fit", "1K samples", () => new OneHotEncoder().fit(strLabels1k_2d));
const ohe = new OneHotEncoder().fit(strLabels500_2d);
run(suite, "OneHotEncoder transform", "500 samples", () => ohe.transform(strLabels500_2d));

// ── OrdinalEncoder ──────────────────────────────────────

run(suite, "OrdinalEncoder fit", "500 samples", () => new OrdinalEncoder().fit(strLabels500_2d));
const oe = new OrdinalEncoder().fit(strLabels500_2d);
run(suite, "OrdinalEncoder transform", "500 samples", () => oe.transform(strLabels500_2d));

// ── LabelBinarizer ──────────────────────────────────────

run(suite, "LabelBinarizer fit", "500 samples", () => new LabelBinarizer().fit(strLabels500));
const lb = new LabelBinarizer().fit(strLabels500);
run(suite, "LabelBinarizer transform", "500 samples", () => lb.transform(strLabels500));

// ── trainTestSplit ──────────────────────────────────────

run(suite, "trainTestSplit", "200x5", () => trainTestSplit(X200, y200, { testSize: 0.2 }));
run(suite, "trainTestSplit", "500x10", () => trainTestSplit(X500, y500, { testSize: 0.2 }));
run(suite, "trainTestSplit", "1Kx10", () => trainTestSplit(X1k, y1k, { testSize: 0.2 }));

// ── KFold ───────────────────────────────────────────────

run(suite, "KFold (k=5)", "500 samples", () => {
	const kf = new KFold({ nSplits: 5 });
	for (const _ of kf.split(X500)) {
		/* iterate */
	}
});
run(suite, "KFold (k=5)", "1K samples", () => {
	const kf = new KFold({ nSplits: 5 });
	for (const _ of kf.split(X1k)) {
		/* iterate */
	}
});
run(suite, "KFold (k=10)", "1K samples", () => {
	const kf = new KFold({ nSplits: 10 });
	for (const _ of kf.split(X1k)) {
		/* iterate */
	}
});

// ── StratifiedKFold ─────────────────────────────────────

run(suite, "StratifiedKFold (k=5)", "500 samples", () => {
	const sf = new StratifiedKFold({ nSplits: 5 });
	for (const _ of sf.split(X500, y500)) {
		/* iterate */
	}
});
run(suite, "StratifiedKFold (k=5)", "1K samples", () => {
	const sf = new StratifiedKFold({ nSplits: 5 });
	for (const _ of sf.split(X1k, y1k)) {
		/* iterate */
	}
});

// ── LeaveOneOut ─────────────────────────────────────────

const Xsmall = makeNumeric(50, 3, 42);
run(suite, "LeaveOneOut", "50 samples", () => {
	const loo = new LeaveOneOut();
	for (const _ of loo.split(Xsmall)) {
		/* iterate */
	}
});

footer(suite, "deepbox-preprocess.json");
